crates/sui-core/src/authority_server.rs

```diff
@@ -540,6 +542,198 @@ impl ValidatorService {
+            if let Some(last_gas_price) = last_gas_price {
+                fp_ensure!(
+                    gas_price == last_gas_price,
+                    SuiError::FailedToSubmitToConsensus(
+                        "SoftBundle: All certificate must have the same gas price".into()
+                    )
+                    .into()
+                );
+            }
+            last_gas_price = Some(gas_price);
+
+            certificate
+                .data()
+                .validity_check(epoch_store.protocol_config())?;
+            total_serialized_size += certificate.data().serialized_size()?;
+        }
+        fp_ensure!(
+            total_serialized_size <= max_serialized_size,
+            SuiError::FailedToSubmitToConsensus("SoftBundle: Bundle size too large".into()).into()
+        );
+
+        let mut responses = Vec::new();
+        responses.resize(certificates.len(), None);
+        let _metrics_guard = metrics.submit_certificate_consensus_latency.start_timer();
+        let mut cont = Vec::new();
+
+        // First pass
+        let mut cert_verify_futures = Vec::new();
+        for (request_idx, certificate) in certificates.into_iter().enumerate() {
+            // 1) Check if cert already executed
+            let tx_digest = *certificate.digest();
+            let already_executed =
+                match state.get_signed_effects_and_maybe_resign(&tx_digest, &epoch_store)? {
+                    Some(_) => true,
+                    None => false,
+                };
+            fp_ensure!(
+                !already_executed,
+                SuiError::FailedToSubmitToConsensus(
+                    "SoftBundle: Rejected bundle with already executed certificate".into()
+                )
+                .into()
+            );
+
+            // 2) Verify the cert.
+            // Check system overload
+            let overload_check_res = state.check_system_overload(
+                &consensus_adapter,
+                certificate.data(),
+                state.check_system_overload_at_signing(),
+            );
+            if let Err(error) = overload_check_res {
+                metrics
+                    .num_rejected_cert_during_overload
+                    .with_label_values(&[error.as_ref()])
+                    .inc();
+                return Err(error.into());
+            }
+
+            let future = epoch_store.signature_verifier.verify_cert(certificate);
+            cert_verify_futures.push(future);
+            cont.push(request_idx);
+        }
+
+        // Second pass
+        let mut verified_certs = Vec::new();
+        let mut certs_idx_to_submit = Vec::new();
+        // 3) All certificates are sent to consensus (at least by some authorities)
+        // For shared objects this will wait until either timeout or we have heard back from consensus.
+        // For owned objects this will return without waiting for certificate to be sequenced
```

Comment:
This comment was probably copy/pasted, but the discussion of owned objects can be removed since soft bundles can only contain shared object txns.